lib/network/internal.nix

```
@@ -0,0 +1,203 @@
+{ lib }:
+let
+  inherit (lib)
+    concatStringsSep
+    elemAt
+    fold
+    foldl
+    genList
+    mod
+    range
+    reverseList
+    toInt
+    ;
+
+  inherit (builtins) toString;
+in
+rec {
+  common = {
+    /**
+      Given a base and exponent, calculates base raised to the exponent.
+
+      # Example
+
+      ```nix
+      pow 2 3
+      => 8
+      ```
+
+      # Type
+
+      ```
+      pow :: Int -> Int -> Int
+      ```
+
+      # Arguments
+
+      - [base] The base.
+      - [exponent] The exponent.
+
+      # Throws
+
+      - If the exponent is less than 0.
+    */
+    pow =
+      base: exponent:
+      if exponent < 0 then
+        throw "lib.network.pow: Exponent cannot be negative."
+      else if exponent == 0 then
+        1
+      else
+        fold (x: y: y * base) base (range 2 exponent);
```

`foldl'` runs in constant space.

Should we be concerned about performance soon?

```suggestion
        foldl' (acc: _: acc * base) base (range 2 exponent)
```

`acc` is typically used for "accumulator"

`_` signifies that we're ignoring that parameter.


Maybe factor out a function for iterated application? I couldn't find one in the existing library. The list stuff here is a bit of a distraction.

You could also fold the `exponent == 0` case into the fold based implementation by starting with an accumulator of 1.

```suggestion
        foldl' (acc: _: acc * base) 1 (range 1 exponent)
```

---

You mean something like this as a function for iterated application?

```nix
repeat = num: times: map (x: if x > 0 then num else 1) (lib.range 1 times);

pow = foldl' mul 1 (repeat base exponent);
```

```nix
repeat 2 4
=> [2 2 2 2]
```

Similarly, with this approach we can get rid of the `exponent == 0` case like you did.

---

> Should we be concerned about performance soon?

Not sure we need to be terribly concerned about performance. For IPv4, the largest this should see is `2^8`. For IPv6, I think we will need another method of encoding addresses.

---

I like Robert's second suggestion the most. I don't think we need to split this off into a separate public function for now, at least until we have more use cases for such a function. I haven't looked myself though, there may already be more use cases, in which case I definitely wouldn't complain about a separate PR for that, but my preference is to not bother for now :)

---

Done!